Sources/_MatchingEngine/Regex/AST/Atom.swift

```diff
@@ -76,6 +76,32 @@ extension AST {
   }
 }
 
+extension AST.Atom {
+  private var _associatedValue: Any? {
+    switch kind {
+    case .char(let v):                  return v
+    case .scalar(let v):                return v
+    case .property(let v):              return v
+    case .escaped(let v):               return v
+    case .keyboardControl(let v):       return v
+    case .keyboardMeta(let v):          return v
+    case .keyboardMetaControl(let v):   return v
+    case .namedCharacter(let v):        return v
+    case .backreference(let v):         return v
+    case .subpattern(let v):            return v
+    case .callout(let v):               return v
+    case .backtrackingDirective(let v): return v
+    case .any:                          return nil
+    case .startOfLine:                  return nil
+    case .endOfLine:                    return nil
+    }
+  }
+
+  func `as`<T>(_ t: T.Type = T.self) -> T? {
+    _associatedValue as? T
+  }
+}
+
```

Comment:
I wonder if it's useful/interesting to formalize this as some kind of `enum-of-structs` protocol.

Reply:
Yeah we could do, though it would mean we'd no longer be able to keep `_associatedValue` private, but maybe that doesn't matter as long as it's underscored